### PR TITLE
Add ContentType detection and Tests for upload ExtraArgs

### PR DESCRIPTION
**Pull request by geoffjukes**
Hi,

`upload_fileobj` fails to detect/set the `content-type` for objects. This can be problematic when displaying images inline, or streaming audio.

These changes refactor the `upload_args` into a method, so that the MimeType for the key can be guessed using the `mimetypes` package, with a fallback to `binary/octet-stream`.

The `mimetypes.guess_type` function is lightweight, and uses the file extension to guess. This is the same method used by other Boto3 methods.

I have also added some tests for the `_upload_args` method, that cover the `acl` and `cache_control` options too.

Hopefully this is acceptable.

### CODE REVIEW

**Comment by geoffjukes** [comment]
@willmcgugan Am I doing something stupid with my pull requests? Should I be re-basing or something? Because I can see all my commits in the PR, even those that have already been merged.

TBH, this is the first project I've really contributed to, and I'm not really sure what I'm doing 😄

**Comment by geoffjukes** [comment]
Any thoughts on this @willmcgugan ? I'd like to do the PR for #36

**Comment by willmcgugan** [comment]
Sorry,  juggling projects lately. I've have a look now. btw feel free to prompt me on twitter if I've forgotten a issue/pr.

File: fs_s3fs/_s3fs.py [hunk 1]
```diff
@@ -353,6 +354,10 @@ def _get_object(self, path, key):
         else:
             return obj
 
+    def _upload_args(self, key):
+        mimetype = mimetypes.guess_type(key)[0] or 'binary/octet-stream'
+        return dict(ContentType=mimetype, **self.upload_args)
```
Comment (willmcgugan) [inline fs_s3fs/_s3fs.py]: `upload_args` is a free dict right? If the user happened to add `ContentType` to `upload_args` this would raise a TypeError.

Comment (geoffjukes) [reply fs_s3fs/_s3fs.py]: @willmcgugan oops, good catch! Fixing now

**Comment by geoffjukes** [comment]
@willmcgugan I added a check for a manually defined `ContentType`. I've tried to avoid 'setting' anything in `self.upload_args` that could change depending on the provided key.

**Comment by geoffjukes** [comment]
@willmcgugan shall I just add the `makedir` change for #36 to this PR too?

**Comment by geoffjukes** [comment]
Typo in the commit message - should have been "fixes #36"

**Comment by willmcgugan** [comment]
Fantastic. Thanks @geoffjukes

**Pull request merged**